modules/programs/firefox.nix

```diff
@@ -21,10 +21,42 @@ let
   profilesPath =
     if isDarwin then "${firefoxConfigPath}/Profiles" else firefoxConfigPath;
 
-  nativeMessagingHostsJoined = pkgs.symlinkJoin {
-    name = "home_ff_nmhs";
-    paths = cfg.package.nativeMessagingHosts or cfg.nativeMessagingHosts;
-  };
+  nativeMessagingHosts = let
+    installNativeMessagingPath = "${mozillaConfigPath}/native-messaging-hosts";
+    sourcePackage = pkgs.symlinkJoin {
+      name = "hm_ff_nativeMessagingHosts";
+      paths = cfg.nativeMessagingHosts ++ [
+        (pkgs.runCommand "hm_ff_emptyNativeMessagingHost" { } ''
+          mkdir -p $out/lib/mozilla/native-messaging-hosts
+        '')
+      ] ++ [ cfg.finalPackage ];
+    };
+    sourceNativeMessagingPath =
+      "${sourcePackage}/lib/mozilla/native-messaging-hosts";
+
+    hostList = filesystem.listFilesRecursive sourceNativeMessagingPath;
+    # Convert list to a list we can convert into an attribute set
+    linkList = (builtins.map (path:
+      let name = last (splitString "/" path);
+      in {
+        # unsafeDiscardStringContext is safe here since we're just constructing a key
+        name = "${installNativeMessagingPath}/${
+            builtins.unsafeDiscardStringContext name
+          }";
+        # We can't use source here without unsafeDiscardStringContext, which we want to avoid
+        value = {
+          text = (builtins.readFile "${sourceNativeMessagingPath}/${name}");
+        };
```

Comment:
This causes IFD. Why does `value.source = "${sourceNativeMessagingPath}/${name}"` not work?

Reply:
For the same reason we have to use builtins.stringDiscardUnsafeContext, filesystem.recursivelisting thingy has a context in the packages. I don't know enough about contexts, but the symlinks could potentially be dead.

See https://discourse.nixos.org/t/not-allowed-to-refer-to-a-store-path-error/5226

Reply:
@Scrumplex I've pushed changes that uses source instead. I'll have to test cleaning all generations and collecting garbage before we know it's stable though.